c10/cuda/CUDAException.cpp

```diff
@@ -26,10 +109,192 @@ void c10_cuda_check_implementation(
   check_message.append("CUDA error: ");
   check_message.append(cudaGetErrorString(err));
   check_message.append(c10::cuda::get_cuda_check_suffix());
+  check_message.append("\n");
+  if (include_device_assertions) {
+    check_message.append(c10_retrieve_device_side_assertion_info());
+  }
 #endif
 
   TORCH_CHECK(false, check_message);
 }
 
+CUDAKernelLaunchRegistry::CUDAKernelLaunchRegistry()
+    : do_all_devices_support_managed_memory(
+          check_if_all_devices_support_managed_memory()) {
+  for (C10_UNUSED const auto _ : c10::irange(get_device_count())) {
+    uvm_assertions.emplace_back(nullptr, uvm_deleter);
+  }
+
+  kernel_launches.resize(max_kernel_launches);
+}
+
+uint32_t CUDAKernelLaunchRegistry::insert(
+    const std::string& launch_filename,
+    const std::string& launch_function,
+    const uint32_t launch_linenum,
+    const std::string& kernel_name,
+    const int32_t stream_id) {
+#ifdef TORCH_USE_CUDA_DSA
+  const std::lock_guard<std::mutex> lock(read_write_mutex);
+  const auto my_gen_number = generation_number++;
+  // TODO: It would probably be good to get a stack trace here so that
+  // we can better indicate which launch caused the failure.
+  kernel_launches[my_gen_number % max_kernel_launches] = {
+      launch_filename,
+      launch_function,
+      launch_linenum,
+      c10::get_backtrace(),
```

Comment:
how slow is getting backtrace? maybe make it opt-in? everything else should be quick. If you want to keep it - move it outside of the mutex

Reply:
I've added a `TORCH_CUDA_DSA_WITH_BACKTRACE` preprocessor gate (unless you think it'd be better to have this be opt-in at runtime?) and moved collection outside of the mutex.

Reply:
hm, it's probably better to make it a runtime opt-in. Environment variable perhaps? Users definitely won't recompile torch to debug things. I'd also make the whole reporting controllable by an env variable in addition to the preprocessor flag. So you can have settings like off/without_stack/with_stack